src/power.jl

```diff
@@ -7,156 +7,251 @@
-    coefhomog = zero(T)
-    for i = 0:kcoef-knull-1
-        aux = x*(kcoef-i)-i
-        @inbounds coefhomog += aux*ac[kcoef-i+1]*coeffs[i+1]
+    coefhomog = zero(eltype(c))
+    for i = 0:k-l0-1
+        aux = r*(k-i) - i
+        @inbounds coefhomog += aux * a[k-i+1] * c[i+1]
     end
-    aux = kcoef - knull*(x+1)
-    @inbounds coefhomog = coefhomog / (aux*ac[knull+1])
+    aux = k - l0*(r+1)
+    @inbounds coefhomog = coefhomog / (aux * a[l0+1])
 
-    coefhomog
+    c[k-l0+1] = coefhomog
+    return nothing
+end
+
+function pow!{S<:Real}(c::TaylorN, a::TaylorN, r::S, k::Int)
+    if k == 0
+        @inbounds c[1] = ( a[1][1] )^r
+        return nothing
+    end
+
+    coefhomog = zero(eltype(c))
+    coefhomog = HomogeneousPolynomial(zero(eltype(c)), k)
+    for i = 0:k-1
+        aux = r*(k-i) - i
+        @inbounds coefhomog += aux * a[k-i+1] * c[i+1]
+    end
+    @inbounds c[k+1] = coefhomog / (k * a[1][1])
+
+    return nothing
 end
 
+
+
 ## Square ##
-function square(a::Taylor1)
-    coeffs = Array{eltype(a)}(a.order+1)
-    coeffs[1] = a[1]^2
-    @inbounds for k = 1:a.order
-        coeffs[k+1] = squareHomogCoef(k, a.coeffs)
+"""
+    square(a::AbstractSeries) --> typeof(a)
+
+Return `a^2`; see [TaylorSeries.sqr!](@ref).
+""" square
+
+for T in (:Taylor1, :TaylorN)
+    @eval function square(a::$T)
+        if $T == Taylor1
+            c = Taylor1(Array{eltype(a)}(a.order+1))
+        else
+            c = TaylorN( zeros(HomogeneousPolynomial{eltype(a)}, a.order) )
+        end
+
+        @inbounds for k = 0:a.order
+            sqr!(c, a, k)
+        end
+        return c
     end
-    Taylor1(coeffs,a.order)
 end
 
+function square(a::HomogeneousPolynomial)
+    T = eltype(a)
+
+    order = 2*a.order
+    order > get_order() && return HomogeneousPolynomial([zero(T)], get_order())
+
+    res = HomogeneousPolynomial([zero(T)], order)
+    sqr!(res, a)
+    return res
+end
+
+
 # Homogeneous coefficients for square
 doc"""
-    squareHomogCoef(kcoef, ac)
+    sqr!(c, a, k::Int) --> nothing
 
-Compute the `k-th` expansion coefficient of $c = a^2$, given by
+Update the `k-th` expansion coefficient `c[k+1]` of `c = a^2`, for
+both `c` and `a` either `Taylor1` or `TaylorN`.
 
-\begin{eqnarray*}
-c_k &=& 2 \sum_{j=0}^{(k-1)/2} a_{k-j} a_j, \text{ if $k$ is odd, or }  \\\\
-c_k &=& 2 \sum_{j=0}^{(k-2)/2} a_{k-j} a_j + (a_{k/2})^2, \text{ if $k$ is even, }
-\end{eqnarray*}
+The coefficients are given by
+
+\\begin{eqnarray*}
+c_k & = & 2 \\sum_{j=0}^{(k-1)/2} a_{k-j} a_j,
+    \\text{ if k is odd, or }  \\\\
+c_k & = & 2 \\sum_{j=0}^{(k-2)/2} a_{k-j} a_j + (a_{k/2})^2,
+    \\text{ if k is even. }
+\\end{eqnarray*}
+
+""" sqr!
+
+for T = (:Taylor1, :TaylorN)
+    @eval function sqr!(c::$T, a::$T, k::Int)
+        if k == 0
+            @inbounds c[1] = a[1]^2
+            return nothing
+        end
 
-with $a$ a `Taylor1` polynomial.
+        if $T == Taylor1
+            @inbounds c[k+1] = zero(c[1])
```

Comment:
I forgot completly this file; I'll implement the same changes as in `functions.jl` here.